flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+		assertEquals(0, spaceAllocator.getTotalSpaceNumber());
+		assertTrue(stateMap.isClosed());
+	}
+
+	/**
+	 * Tests that remove states physically during sync part of snapshot.
+	 */
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testPhysicallyRemoveDuringSyncPartOfSnapshot() throws IOException {
+		TypeSerializer<Integer> keySerializer = IntSerializer.INSTANCE;
+		TypeSerializer<Long> namespaceSerializer = LongSerializer.INSTANCE;
+		TypeSerializer<String> stateSerializer = StringSerializer.INSTANCE;
+		// set logicalRemovedKeysRatio to 0 so that all logically removed states will be deleted when snapshot
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap = new CopyOnWriteSkipListStateMap<>(
+			keySerializer,
+			namespaceSerializer,
+			stateSerializer,
+			spaceAllocator,
+			0,
+			0.0f);
+
+		// map to store expected states, namespace -> key -> state
+		Map<Long, Map<Integer, String>> referenceStates = new HashMap<>();
+		int totalStateSize = 0;
+		int totalSizeIncludingLogicallyRemovedStates = 0;
+
+		// put some states
+		for (int i = 1; i <= 100; i++) {
+			totalStateSize++;
+			totalSizeIncludingLogicallyRemovedStates++;
+			stateMap.put(i, (long) i, String.valueOf(i));
+			addToReferenceState(referenceStates, i, (long) i, String.valueOf(i));
+		}
+		verifyState(referenceStates, stateMap);
+
+		assertEquals(totalStateSize * 2, spaceAllocator.getTotalSpaceNumber());
+
+		Map<Long, Map<Integer, String>> expectedSnapshot1 = snapshotReferenceStates(referenceStates);
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot1 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		// remove all states logically
+		for (int i = 1; i <= 100; i++) {
+			totalStateSize--;
+			stateMap.remove(i, (long) i);
+			removeFromReferenceState(referenceStates, i, (long) i);
+		}
+		assertEquals(totalSizeIncludingLogicallyRemovedStates * 3, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(0, totalStateSize);
+		assertEquals(totalStateSize, stateMap.size());
+		assertEquals(totalSizeIncludingLogicallyRemovedStates, stateMap.totalSize());
+		assertEquals(totalSizeIncludingLogicallyRemovedStates, stateMap.getLogicallyRemovedNodes().size());
+		verifyState(referenceStates, stateMap);
+
+		verifySnapshotWithoutTransform(
+			expectedSnapshot1, snapshot1, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot1.release();
+
+		// no spaces should be free
+		assertEquals(totalSizeIncludingLogicallyRemovedStates * 3, spaceAllocator.getTotalSpaceNumber());
+		verifyState(referenceStates, stateMap);
+
+		Map<Long, Map<Integer, String>> expectedSnapshot2 = snapshotReferenceStates(referenceStates);
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot2 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		// all state should be removed physically
+		totalSizeIncludingLogicallyRemovedStates = 0;
```

Comment:
Instead of setting new values I would suggest to introduce new names which give meaning to the value. E.g. `totalSizeIncludingLogicallyRemovedStatesAfterSecondSnapshot` or so.